### PR TITLE
[WIP] Layout algorithm

### PR DESCRIPTION
**Pull request by zamotany**
TODO:
- [ ] layers support
- [ ] documentation
- [ ] profiling (**help wanted**) + optimisation
- [x] tests
- [ ] boundaries and guards

`yarn start box`
Use W/A/S/D to move the box.

### CODE REVIEW

File: src/nodes/ChunkNode.js [hunk 1]
```diff
@@ -1,24 +1,44 @@
...
     this.container = container;
     this.props = props;
   }
 
+  setParentsOffset({ x, y }: { x: number, y: number }) {
+    this.parentsOffset.x = x;
+    this.parentsOffset.y = y;
+  }
+
+  getParentsOffset() {
```
Comment (satya164) [inline src/nodes/ChunkNode.js]: this is not parent offset, but the calculated offset of current element, right?

File: src/utils/Layout.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+function getLines(text) {
+  let lines = text.split('\n');
+
+  if (lines.length === 1) {
+    return lines;
+  }
+
+  const endsWithNewLine = text.endsWith('\n');
+  if (endsWithNewLine) {
+    lines = lines.slice(0, lines.length - 1);
```
Comment (satya164) [inline src/utils/Layout.js]: nit: same as `lines.slice(0, -1)`

File: src/utils/Layout.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    this.lastFilledLine = toLine;
+  }
+
+  build() {
+    // @TODO: support layers (z-index)
+
+    // Assumption: both x and y are zero-based indexes
+    this.elements.forEach(element => {
+      const contentLines = getLines(element.text);
```
Comment (satya164) [inline src/utils/Layout.js]: Some comments here will be nice. Took me a while.

File: src/utils/getPositionPros.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/* @flow */
```
Comment (satya164) [inline src/utils/getPositionPros.js]: nit: typo in filename and usage places

**Comment by zamotany** [comment]
I need a feedback regarding this idea: can we force every component to be positioned absolutely and create component for relative positioning that will iterate over children and assign absolute positions under the hood?

**Pull request merged**